src/core/lib/security/credentials/s2a/grpc_s2a_credentials_options.cc

```diff
@@ -22,43 +22,89 @@
 
 #include <grpc/support/alloc.h>
 #include <grpc/support/log.h>
+#include <grpc/support/string_util.h>
+#include <string.h>
+
+#include <iostream>
 
 namespace experimental {
 
-grpc_s2a_credentials_options* grpc_s2a_credentials_options_copy(
-    const grpc_s2a_credentials_options* options) {
-  if (options != nullptr && options->vtable != nullptr &&
-      options->vtable->copy != nullptr) {
-    return options->vtable->copy(options);
+grpc_s2a_credentials_options::~grpc_s2a_credentials_options() {
+  if (handshaker_service_url_ != nullptr) {
+    gpr_free(handshaker_service_url_);
+  }
+  for (auto service_account : target_service_account_list_) {
+    gpr_free(service_account);
   }
-  gpr_log(GPR_ERROR, "Invalid arguments to grpc_s2a_credentials_options_copy.");
-  return nullptr;
 }
 
-void grpc_s2a_credentials_options_destroy(
-    grpc_s2a_credentials_options* options) {
-  if (options != nullptr) {
-    if (options->vtable != nullptr && options->vtable->destruct != nullptr) {
-      options->vtable->destruct(options);
+void grpc_s2a_credentials_options::set_handshaker_service_url(
+    const char* handshaker_service_url) {
+  if (handshaker_service_url_ != nullptr) {
+    gpr_free(handshaker_service_url_);
+  }
+  handshaker_service_url_ = gpr_strdup(handshaker_service_url);
+}
+
+void grpc_s2a_credentials_options::add_supported_ciphersuite(
+    uint16_t ciphersuite) {
+  supported_ciphersuites_.push_back(ciphersuite);
+}
+
+void grpc_s2a_credentials_options::add_target_service_account(
+    const char* target_service_account) {
+  char* service_account = gpr_strdup(target_service_account);
+  target_service_account_list_.push_back(service_account);
+}
+
+grpc_s2a_credentials_options* grpc_s2a_credentials_options::copy() const {
+  grpc_s2a_credentials_options* new_options =
+      new grpc_s2a_credentials_options();
+  if (handshaker_service_url_ != nullptr) {
+    new_options->set_handshaker_service_url(handshaker_service_url_);
+  }
+  for (auto ciphersuite : supported_ciphersuites_) {
+    new_options->add_supported_ciphersuite(ciphersuite);
+  }
+  for (auto service_account : target_service_account_list_) {
+    new_options->add_target_service_account(service_account);
+  }
+  return new_options;
+}
+
+bool grpc_s2a_credentials_options::check_fields(
```

Comment:
Is this API used anywhere other than tests? If not, can we make it a static method instead of class member function?

Reply:
Yes, it is used for testing purposes. I've added a comment to that effect.